lib/puppetserver/ca/action/generate.rb

```diff
@@ -168,25 +212,77 @@ def sign_intermediate(ca_key, ca_cert, csr, valid_until, signing_digest)
             extension = ef.create_extension(*ext)
             cert.add_extension(extension)
           end
+
           cert.sign(ca_key, signing_digest)
 
           cert
         end
 
-        def parse(args)
-          results = {}
-          parser = self.class.parser(results)
+        def sign_master_cert(int_key, int_cert, csr, valid_until, signing_digest, subject_alt_names)
+          cert = OpenSSL::X509::Certificate.new
+          cert.public_key = csr.public_key
+          cert.subject = csr.subject
+          cert.issuer = int_cert.subject
+          cert.version = 2
+          cert.serial = 1
+          cert.not_before = CERT_VALID_FROM
+          cert.not_after = valid_until
 
-          errors = CliParsing.parse_with_errors(parser, args)
+          ef = extension_factory_for(int_cert, cert)
+          MASTER_EXTENSIONS.each do |ext|
+            extension = ef.create_extension(*ext)
+            cert.add_extension(extension)
+          end
 
-          errors_were_handled = CliParsing.handle_errors(@logger, errors, parser.help)
+          if !subject_alt_names.empty?
+            alt_names_ext = ef.create_extension("subjectAltName", subject_alt_names, false)
+            cert.add_extension(alt_names_ext)
+          end
 
-          exit_code = errors_were_handled ? 1 : nil
+          cert.sign(int_key, signing_digest)
+          cert
+        end
 
+        # Given any alt names specified on the command line and any loaded from
+        # settings, choose the names with the highest precedence and process them.
+        # Precedence:
+        # 1) Specified on the CLI
+        # 2) Specified in settings
+        # 3) Defaults
+        def choose_alt_names(cli_alt_names, settings_alt_names, certname)
+          if !cli_alt_names.empty?
+            sans = cli_alt_names
+          elsif !settings_alt_names.empty?
+            sans = settings_alt_names
+          else
+            sans = "puppet, #{certname}"
```

Comment:
Should this default live in PuppetConfig?

Reply:
Oh yeah, that would make sense.